src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m

```diff
@@ -0,0 +1,133 @@
+//----------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//----------------------------------------------------------------
+
+#import "MSInstallationTemplate.h"
+#import "MSTagHelper.h"
+
+@interface MSInstallationTemplate ()
+
+@property(nonatomic, copy) NSSet<NSString *> *tags;
+@property(nonatomic, copy) NSDictionary<NSString *, NSString *> *headers;
+
+@end
+
+@implementation MSInstallationTemplate
+
+@synthesize body;
+@synthesize tags;
+@synthesize headers;
+
+- (instancetype)init {
+    if (self = [super init]) {
+        tags = [NSSet new];
+        headers = [NSDictionary new];
+    }
+    return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder {
+    if (self = [super init]) {
+        body = [coder decodeObjectForKey:@"body"] ?: @"";
+        tags = [coder decodeObjectForKey:@"tags"];
+        headers = [coder decodeObjectForKey:@"headers"];
+    }
+
+    return self;
+}
+
+#pragma mark Tags
+
+- (BOOL)addTag:(NSString *)tag {
+    return [self addTags:[NSArray arrayWithObject:tag]];
+}
+
+- (BOOL)addTags:(NSArray<NSString *> *)tagsToAdd {
+    NSMutableSet *tmpTags = [NSMutableSet setWithSet:self.tags];
+
+    for (NSString *tag in tagsToAdd) {
```

Comment:
This doesn't _really_ track changes to the underlying set. If we are going to return `BOOL`, then we should try to make it accurate. That probably means we need to iterate over the tags to add, check if the tag exists, add it, and mark the set as "modified".

Reply:
Adding change tracking via `isDirty` and KVO